### PR TITLE
PAM: support the authentication facility

### PR DESCRIPTION
**Pull request by valpackett**
### Motivation and Context

In the system I'm working on, I'd like to exclusively use the ZFS dataset passphrase for user authentication, in order to avoid extra sources of persistent state (in this case `/etc/shadow` style files). Seems like this could be useful for others too so I'd like to upstream it.

### Description

Implement the `pam_sm_authenticate` method, using the `noop` argument of `lzc_load_key` to do a passphrase check without actually loading the key.

This allows using ZFS as the source of truth for user passwords, without storing any password hashes in `/etc` or using other PAM modules.

#### Notes for Other PRs

If #13050 is to land after this, it should either extract the "already mounted" check up to `pam_sm_open_session` or skip that check in noop mode.

### How Has This Been Tested?

Adding these to `/etc/pam.d/su`

```
auth            sufficient      pam_zfs_key.so          debug homes=testpool prop_mountpoint

session         optional        pam_zfs_key.so          debug homes=testpool prop_mountpoint
```

and adding a test user with a home directory pointing to an encrypted dataset (using `vipw`),

then verifying that `su uzer` correctly checks the passphrase but doesn't mount anything, while `su -l uzer` (`-l` for "Simulate a full login" that does open a session) does also mount and unmount the home dataset.

### Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Performance enhancement (non-breaking change which improves efficiency)
- [ ] Code cleanup (non-breaking change which makes code smaller or more readable)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Library ABI change (libzfs, libzfs\_core, libnvpair, libuutil and libzfsbootenv)
- [ ] Documentation (a change to man pages or other documentation)

### Checklist:

- [x] My code follows the OpenZFS [code style requirements](https://github.com/openzfs/zfs/blob/master/.github/CONTRIBUTING.md#coding-conventions).
- [ ] I have updated the documentation accordingly.
- [x] I have read the [**contributing** document](https://github.com/openzfs/zfs/blob/master/.github/CONTRIBUTING.md).
- [ ] I have added [tests](https://github.com/openzfs/zfs/tree/master/tests) to cover my changes.
- [ ] I have run the ZFS Test Suite with this change applied.
- [x] All commit messages are properly formatted and contain [`Signed-off-by`](https://github.com/openzfs/zfs/blob/master/.github/CONTRIBUTING.md#signed-off-by).

### CODE REVIEW

File: contrib/pam_zfs_key/pam_zfs_key.c [hunk 1]
```diff
@@ -463,7 +467,7 @@ zfs_key_config_load(pam_handle_t *pamh, zfs_key_config_t *config,
 	if (!entry) {
 		free(config->runstatedir);
 		free(config->homes_prefix);
-		return (-1);
+		return (-2);
```
Comment (valpackett) [inline contrib/pam_zfs_key/pam_zfs_key.c:470]: bikeshedding: probably a good idea to make return codes here nicer, but how exactly? enum or defines? at the top of the file?

**Comment by behlendorf** [comment]
@felixdoerre would you mind taking a look at this PR.

File: contrib/pam_zfs_key/pam_zfs_key.c [hunk 2]
```diff
@@ -646,10 +650,44 @@ pam_sm_authenticate(pam_handle_t *pamh, int flags,
...
+	}
+	if (pam_zfs_init(pamh) != 0) {
+		zfs_key_config_free(&config);
+		return (PAM_SERVICE_ERR);
+	}
+	char *dataset = zfs_key_config_get_dataset(&config);
+	if (!dataset) {
+		pam_zfs_free();
+		zfs_key_config_free(&config);
+		return (PAM_IGNORE);
```
[the remark this answers is not in the document]
Comment (valpackett) [reply contrib/pam_zfs_key/pam_zfs_key.c:680]: I thiiink `PAM_IGNORE` does make sense exactly here in the "no dataset" case, as this is a "not good password, not bad password, but we're out" situation.

[pam_google_authenticator returns it](https://github.com/google/google-authenticator-libpam/blob/fdd8289be83c6154b2fb8d961dbb8a163a9daad4/src/pam_google_authenticator.c#L2112) if it doesn't have a secret for a user… and the `nullok` flag is set. Right, this probably should be configurable indeed.

File: contrib/pam_zfs_key/pam_zfs_key.c (hunk 1 above)
[the remark this answers is not in the document]
Comment (valpackett) [reply contrib/pam_zfs_key/pam_zfs_key.c:470]: Oh. Brilliant!

File: contrib/pam_zfs_key/pam_zfs_key.c [hunk 3]
```diff
@@ -646,10 +650,44 @@ pam_sm_authenticate(pam_handle_t *pamh, int flags,
 {
 	(void) flags, (void) argc, (void) argv;
 
-	if (pw_fetch_lazy(pamh) == NULL) {
-		return (PAM_AUTH_ERR);
+	if (geteuid() != 0) {
+		pam_syslog(pamh, LOG_ERR,
+		    "Cannot zfs_mount when not being root.");
+		return (PAM_IGNORE);
```
[the remark this answers is not in the document]
Comment (valpackett) [reply contrib/pam_zfs_key/pam_zfs_key.c:656]: Right, let's stick to `PAM_SERVICE_ERR` here I guess

**Review by behlendorf — APPROVED** [review]

**Pull request merged**